Fix threading bugs

bcfriesen opened the pull request:
There are several race conditions in Nyx, many related to updating atomic data inside OpenMP parallel regions. This series of commits removes most of them.

bcfriesen committed:
EOS: add BIND(C) attribute to interp_to_this_z() function, and add C++ function prototype

This function modifies public variables stored in modules, so we want to have
control over which threads execute this function; if multiple threads execute
it simultaneously, then there is a race condition. The OpenMP semantics should
be in the C++ code, not Fortran, so we will move calls to this function out of
existing Fortran functions and into C++ code instead.

bcfriesen committed:
EOS: move Fortran calls to interp_to_this_z() to C++

The C++ calls to interp_to_this_z() are all done in serial regions, so the race
condition of calling interp_to_this_z() from a threaded region is gone.

bcfriesen committed:
Nyx::just_the_hydro(): add OpenMP reduction construct to e_added,ke_added

This fixes a race condition, which is that they were added together without
atomics or using thread-private copies.

bcfriesen committed:
Nyx::strang_second_step(): add OpenMP reduction construct to min_iter, max_iter

This fixes a race condition in the min/max comparisons of these variables.

bcfriesen committed:
EOS, HeatCool: fix naming mixups between "interp_to_this_z()" and "fort_interp_to_this_z()"

The convention is to add the "fort_" prefix to Fortran functions if they are
called from C++. I messed this up at some point, so now it's all consistent.

bcfriesen committed:
EOS, HeatCool: move z_vode update to fort_interp_to_this_z()

This removes a race condition in updating z_vode.

WeiqunZhang commented on Source/Nyx_hydro.cpp:
```diff
@@ -129,11 +129,13 @@ Nyx::just_the_hydro (Real time,
        {
        FArrayBox flux[BL_SPACEDIM], u_gdnv[BL_SPACEDIM];
        Real cflLoc = -1.e+200;
+       e_added = 0.0;
```
My understanding is the initialization should be outside omp parallel because they are shared.

bcfriesen replied:
I wrote it that way because I did not understand initialization rules for reduction variables. However, Section 2.15.3.6 of the OpenMP 4.5 spec (as well as Table 2.7) says that the default initializer for the `+` reduction is 0, which is what we want. So the code that I wrote is not technically wrong, but rather is redundant. I will remove the initializations from inside the parallel region.

bcfriesen committed:
Nyx::just_the_hydro(): remove manual initialization of reduction variables

In commit f3bdfe4 ("Nyx::just_the_hydro(): add OpenMP reduction construct to
e_added,ke_added") I moved initialization of the reduction variables "e_added"
and "ke_added" to inside the OpenMP parallel region. However, while this code
is not wrong, it is unnecessary because the OpenMP spec requires that the
default initializer for the "+" reduction is zero, which is what this code
already does by hand. So we can remove it.

bcfriesen commented:
@WeiqunZhang commit 090197c should make this correct.

bcfriesen commented:
Oops, forgot about that case. OK I will fix that.

bcfriesen committed:
Nyx::just_the_hydro(): restore manual initialization of reduction variables

If OpenMP is disabled, the variables "e_added" and "ke_added" still should be
initialized to zero, so initialize them outside of the OpenMP parallel region.

bcfriesen commented:
Oops one sec, copy-paste error in the latest commit.
